service/history/historyEngine.go

```diff
@@ -502,7 +502,7 @@ func (e *historyEngineImpl) StartWorkflowExecution(
 		startRequest,
 	)
 	if err != nil {
-		return nil, serviceerror.NewUnavailable("Failed to add workflow execution started event.")
+		return nil, err
```

Comment:
How do you decide when to directly return the error and when to wrap it in serviceerror.NewInternal?

Reply:
not 100% clear on this, but roughly my current view:

error generated within our logic, e.g. business logic and persistence wrapper should be just returned (assuming caller does not need to do special thing, e.g. workflow ID reuse policy)
currently there are soo many place where logic simply just return a new error or wrap the existing error without too much meaning. (i can be totally wrong)

if we desire stack trace, log.error should be add where error is generated
if we want additional info about this public API (gRPC) involved, we can additionally log the API within network interceptor.

for these specific cases, e.g. history engine creating new errors: mutable state already generated the internal error. i guess it would make the code cleaner (to just return)